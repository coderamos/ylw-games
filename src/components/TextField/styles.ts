import styled, { css } from 'styled-components';

export const TextFieldWrapper = styled.div``;

export const InputWrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    background: ${theme.colors.grayLight};
    border-radius: 0.2rem;
    padding: 0 ${theme.spacings.xxsmall};
    border: 0.2rem solid;
    border-color: ${theme.colors.primary};

    &:focus-within {
      box-shadow: 0 0 0.5rem ${theme.colors.primary};
    }
  `}
`;

export const Input = styled.input`
  ${({ theme }) => css`
    color: ${theme.colors.black};
    font-family: ${theme.font.family};
    font-size: ${theme.font.sizes.medium};
    padding: ${theme.spacings.xxsmall} 0;
    background: transparent;
    border: none;
    outline: none;
    width: 100%;
  `}
`;

export const Label = styled.label`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.black};
    cursor: pointer;
  `}
`;
