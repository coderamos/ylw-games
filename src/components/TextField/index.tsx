import { InputHTMLAttributes, useState } from 'react';

import * as s from './styles';

export type TextFieldProps = {
  label?: string;
  initialValue?: string;
  icon?: React.ReactNode;
  iconPosition?: 'left' | 'right';
  disabled?: boolean;
  error?: string;
  onInput?: (value: string) => void;
} & InputHTMLAttributes<HTMLInputElement>;

const TextField = ({
  disabled = false,
  error,
  icon,
  iconPosition = 'left',
  initialValue = '',
  label,
  name,
  onInput,
  ...props
}: TextFieldProps) => {
  const [value, setValue] = useState(initialValue);
  const onChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.currentTarget.value;
    setValue(newValue);

    !!onInput && onInput(newValue);
  };

  return (
    <s.Wrapper disabled={disabled} error={!!error}>
      {!!label && <s.Label htmlFor={name}>{label}</s.Label>}
      <s.InputWrapper>
        {!!icon && (
          <s.IconWrapper iconPosition={iconPosition}>{icon}</s.IconWrapper>
        )}
        <s.Input
          type="text"
          onChange={onChange}
          value={value}
          iconPosition={iconPosition}
          disabled={disabled}
          name={name}
          {...(label ? { id: name } : {})}
          {...props}
        />
      </s.InputWrapper>
      {!!error && <s.Error>{error}</s.Error>}
    </s.Wrapper>
  );
};

export default TextField;
