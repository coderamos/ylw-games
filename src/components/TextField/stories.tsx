import { Story, Meta } from '@storybook/react/types-6-0';

import { MailIcon } from 'assets/icons';

import TextField, { TextFieldProps } from '.';

export default {
  title: 'Base/Form/TextField',
  component: TextField,
  args: {
    label: 'e-mail',
    labelFor: 'email',
    id: 'email',
    initialValue: '',
    placeholder: 'diego@yellow.com',
    icon: <MailIcon />
  },
  argTypes: {
    onInput: { action: 'changed' },
    icon: { type: '' }
  },
  parameters: {
    jest: ['TextField.test.tsx']
  }
} as Meta;

export const TextFieldDefault: Story<TextFieldProps> = args => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField {...args} />
  </div>
);

export const TextFieldWithError: Story<TextFieldProps> = args => (
  <div style={{ maxWidth: 300, padding: 15 }}>
    <TextField error="...something is wrong" {...args} />
  </div>
);
