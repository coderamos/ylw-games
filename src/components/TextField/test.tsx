import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { renderWithTheme } from 'utils/tests/helpers';
import { Email } from '@styled-icons/material-outlined';

import TextField from '.';

describe('<TextField />', () => {
  it('should render with label', () => {
    renderWithTheme(<TextField label="label" labelFor="field" id="field" />);

    expect(screen.getByLabelText('label')).toBeInTheDocument();
  });

  it('should render without Label', () => {
    renderWithTheme(<TextField />);

    expect(screen.queryByLabelText('label')).not.toBeInTheDocument();
  });

  it('should render with placeholder', () => {
    renderWithTheme(<TextField placeholder="some placeholder" />);

    expect(screen.getByPlaceholderText('some placeholder')).toBeInTheDocument();
  });

  it('should change value when typing', async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="textField"
        labelFor="textField"
        id="textField"
      />
    );

    const input = screen.getByRole('textbox');
    const text = 'this is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).toHaveValue(text);
      expect(onInput).toHaveBeenCalledTimes(text.length);
    });
    expect(onInput).toHaveBeenCalledWith(text);
  });

  it('should accessible by tab key', () => {
    renderWithTheme(
      <TextField label="textField" labelFor="textField" id="textField" />
    );

    const input = screen.getByLabelText('textField');
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(input).toHaveFocus();
  });

  it('should renders with Icon', () => {
    renderWithTheme(<TextField icon={<Email data-testid="icon" />} />);

    expect(screen.getByTestId('icon')).toBeInTheDocument();
  });

  it('shoud renders with Icon on the right side', () => {
    renderWithTheme(
      <TextField icon={<Email data-testid="icon" />} iconPosition="right" />
    );

    expect(screen.getByTestId('icon').parentElement).toHaveStyle({ order: 1 });
  });

  it('shoul not change values when disabled', async () => {
    const onInput = jest.fn();
    renderWithTheme(
      <TextField
        onInput={onInput}
        label="textField"
        labelFor="textField"
        id="textField"
        disabled
      />
    );

    const input = screen.getByRole('textbox');
    expect(input).toBeDisabled();

    const text = 'this is my new text';
    userEvent.type(input, text);

    await waitFor(() => {
      expect(input).not.toHaveValue(text);
    });
    expect(onInput).not.toHaveBeenCalled();
  });

  it('should render with error', () => {
    const { container } = renderWithTheme(
      <TextField
        icon={<Email data-testid="icon" />}
        label="textField"
        labelFor="textField"
        error="error message"
      />
    );

    expect(screen.getByText('error message')).toBeInTheDocument();
    expect(screen.getByText('error message')).toHaveStyle({ color: '#E72176' });
    expect(container.firstChild).toMatchSnapshot();
  });
});
