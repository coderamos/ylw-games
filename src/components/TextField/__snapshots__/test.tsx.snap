// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> should render with error 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 .c3 {
  border-color: #E72176;
}

.c0 .c5,
.c0 .c1 {
  color: #E72176;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 0.8rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #FFCB00;
}

.c8 {
  color: #000000;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 0.8rem;
  background: transparent;
  border: none;
  outline: none;
  width: 100%;
}

.c2 {
  font-size: 1.4rem;
  color: #000000;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 2.2rem;
  color: #808080;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 100%;
}

.c9 {
  color: #E72176;
  font-size: 1.2rem;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="textField"
  >
    textField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        data-testid="icon"
        fill="none"
        focusable="false"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z"
        />
        <polyline
          points="22,6 12,13 2,6"
        />
      </svg>
    </div>
    <input
      class="c8"
      id="textField"
      name="textField"
      type="text"
      value=""
    />
  </div>
  <p
    class="c9"
  >
    error message
  </p>
</div>
`;
